src/BD_sensor.c

```diff
@@ -0,0 +1,541 @@
+    }
+    BD_i2c_stop();
+    if (BD_Check_OddEven(b) && (b & 0x3FF) < 1020){
+        b = (b & 0x3FF);
+        if(BD_read_flag==1018&&(b<1000)){
+            b = b - z_ofset;
+            if(b>1024)
+                b=0;
+        }
+
+    }
+    else
+        b=1024;
+
+#if 0
+    sda_gpio_in=gpio_in_setup(sda_pin, 1);
+    b=0;
+    b=gpio_in_read(sda_gpio_in)*300+1;
+#endif
+    BD_read_lock=0;
+    return b;
+}
+
+void BD_i2c_write(unsigned int addr)
+{
+    BD_I2C_start();
+    //// write
+    BD_setLow(sda_gpio);
+    BD_setHigh(scl_gpio);
+    ndelay_bd(delay_m*2);
+    BD_setLow(scl_gpio);
+    addr=BD_Add_OddEven(addr);
+    ///write address
+    ndelay_bd(delay_m);
+    for (int i=10; i >=0; i--)
+    {
+        if ((addr>>i)&0x01)
+        {
+            BD_setHigh(sda_gpio);
+        }
+        else
+            BD_setLow(sda_gpio);
+        ndelay_bd(delay_m);
+        BD_setHigh(scl_gpio);
+        ndelay_bd(delay_m*2);
+        BD_setLow(scl_gpio);
+        ndelay_bd(delay_m);
+    }
+    BD_i2c_stop();
+}
+
+uint32_t INT_to_String(uint32_t BD_z1,uint8_t*data)
+{
+    uint32_t BD_z=BD_z1;
+    uint32_t len=0,j=0;
+    if(BD_z>=1000)
+    {
+        j=BD_z/1000;
+        data[len++] = '0'+j;
+        BD_z-=1000*j;
+        data[len]='0';
+        data[len+1]='0';
+        data[len+2]='0';
+    }
+    if(BD_z>=100)
+    {
+        j=BD_z/100;
+        data[len++] = '0'+j;
+        BD_z-=100*j;
+        data[len]='0';
+        data[len+1]='0';
+    }
+    else if(data[len])
+        len++;
+    if(BD_z>=10)
+    {
+        j=BD_z/10;
+        data[len++] = '0'+j;
+        BD_z-=10*j;
+        data[len]='0';
+    }
+    else if(data[len])
+        len++;
+    j=BD_z;
+    data[len++] = '0'+j;
+    data[len]=0;
+    return len;
+}
+
+//for gcode command
+void
+command_I2C_BD_receive(uint32_t *args)
+{
+    uint8_t oid = args[0];
+    uint8_t data[8];
+    uint16_t BD_z;
+
+    //if(BD_read_flag==1018)
+    //    BD_z=BD_Data;
+    //else
+    BD_z=BD_i2c_read();//BD_Data;
+    BD_Data=BD_z;
+    memset(data,0,8);
+    uint32_t len=0,j=0;
+
+///////////same as function itoa()
+    if(BD_z>=1000)
+    {
+        j=BD_z/1000;
+        data[len++] = '0'+j;
+        BD_z-=1000*j;
+        data[len]='0';
+        data[len+1]='0';
+        data[len+2]='0';
+    }
+    if(BD_z>=100)
+    {
+        j=BD_z/100;
+        data[len++] = '0'+j;
+        BD_z-=100*j;
+        data[len]='0';
+        data[len+1]='0';
+    }
+    else if(data[len])
+        len++;
+    if(BD_z>=10)
+    {
+        j=BD_z/10;
+        data[len++] = '0'+j;
+        BD_z-=10*j;
+        data[len]='0';
+    }
+    else if(data[len])
+        len++;
+    j=BD_z;
+    data[len++] = '0'+j;
+
+    sendf("I2C_BD_receive_response oid=%c response=%*s", oid,len,data);
+}
+
+DECL_COMMAND(command_I2C_BD_receive, "I2C_BD_receive oid=%c data=%*s");
+
+
+void
+command_I2C_BD_send(uint32_t *args)
+{
+    int addr=atoi((char *)args[2]);
+    BD_read_flag=addr;
+    if(addr==1015)
```

Comment:
maybe all these const's should be an enum? 1015, 1023, 1024, .. so many number all over the place :))